xcode/Subconscious/Shared/Components/AddressBook/AddressBookModel.swift

```diff
@@ -61,25 +166,111 @@ struct AddressBookModel: ModelProtocol {
                 }
             }
             
-            return Update(state: model)
+            // TODO: this doesn't capture errors, it just returns an empty list
+            // also, it's scary
+            let fx: Fx<AddressBookAction> =
+                environment.data.listPetnamesAsync()
+                // get the DID for each petname
+                .flatMap { follows in
+                    return Publishers.MergeMany(follows.map { f in
+                        return environment.data.getPetnameAsync(petname: f)
+                            .map { did -> AddressBookEntry? in
+                                guard let did = did else { return nil }
+                                return AddressBookEntry(pfp: Image("dog-pfp"), petname: f, did: did)
+                            }
+                            .compactMap { $0 }
+                    })
+                    .collect()
+                    .eraseToAnyPublisher()
+                }
+                .catch({ error in
+                    Just([])
+                })
+                .map({ follows in
+                    AddressBookAction.populate(follows)
+                })
+                .eraseToAnyPublisher()
```

Comment:
We'll be going via a SQLite table instead of directly populating the model, but this logic has to exist somewhere in the pipeline to produce the data format we're after.